### PR TITLE
GH-2291: As a developer I want to use node's native ES module support and get rid of esm

### PR DESCRIPTION
**Pull request by mor-n4**
See #2291.

### CODE REVIEW

**Comment by mor-n4** [comment]
@mmews-n4 I'm not sure if this is ready to merge, yet. In any case, tests are still missing. But I also have to check a few other things. But it should be ready for a first review.

File: version.json [hunk 1]
```diff
@@ -1,5 +1,5 @@
 {
 	"major": "0",
-	"minor": "30",
-	"tag": ""
+	"minor": "31",
+	"tag": "next"
```
Comment (mmews-n4) [inline version.json:4]: is this change intentional?

**Review by mmews-n4 — APPROVED** [review]
👍 great to finally get that on master!

File: plugins/org.eclipse.n4js.transpiler.es/src/org/eclipse/n4js/transpiler/es/transform/CommonJsImportsTransformation.xtend [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+			.filter[!bare] // ignore bare imports
+			.index[importDecl | state.info.getImportedModule(importDecl)];
+
+		val varStmnts = newArrayList;
+		for (targetModule : importDeclsPerImportedModule.keySet) {
+			varStmnts += transformImportDecl(targetModule, importDeclsPerImportedModule.get(targetModule));
+		}
+
+		val lastImportDecl = state.im.scriptElements.filter(ImportDeclaration).last();
+		insertAfter(lastImportDecl, varStmnts);
```
[the remark this answers is not in the document]
Comment (mor-n4) [reply plugins/org.eclipse.n4js.transpiler.es/src/org/eclipse/n4js/transpiler/es/transform/CommonJsImportsTransformation.xtend:81]: No. A single import that requires rewriting is rewritten into a (changed) import + a new `const`. The changed import is added in place, only the additional const is added after the last import in the file.

File: plugins/org.eclipse.n4js/src/org/eclipse/n4js/N4JSGlobals.java [hunk 3]
```diff
@@ -119,17 +124,39 @@
...
 	/**
-	 * Unmodifiable list containing {@link #N4JSD_FILE_EXTENSION},
-	 * {@link #N4JS_FILE_EXTENSION},{@link #N4JSX_FILE_EXTENSION}, {@link #JS_FILE_EXTENSION},
-	 * {@link #JSX_FILE_EXTENSION}.
+	 * All standard plain-JS file extensions.
+	 * <p>
+	 * Unmodifiable list containing {@link #N4JS_FILE_EXTENSION}, {@link #N4JSD_FILE_EXTENSION},
+	 * {@link #N4JSX_FILE_EXTENSION}, and {@link #JSX_FILE_EXTENSION}.
+	 */
+	public static final Set<String> ALL_JS_FILE_EXTENSIONS = ImmutableSet.of(
```
[the remark this answers is not in the document]
Comment (mor-n4) [reply plugins/org.eclipse.n4js/src/org/eclipse/n4js/N4JSGlobals.java:132]: oops, done!

File: testhelpers/org.eclipse.n4js.cli.tests.helper/src/org/eclipse/n4js/cli/helper/N4jscTestLanguageClient.java [hunk 4]
```diff
@@ -94,7 +94,10 @@ public void afterGenerate(URI source, URI generated) {
 		}
 
 		String fileName = generated.lastSegment();
-		if (!fileName.endsWith(N4JSGlobals.JS_FILE_EXTENSION) && !fileName.endsWith(N4JSGlobals.JSX_FILE_EXTENSION)) {
+		if (!fileName.endsWith("." + N4JSGlobals.JS_FILE_EXTENSION)
+				&& !fileName.endsWith("." + N4JSGlobals.CJS_FILE_EXTENSION)
+				&& !fileName.endsWith("." + N4JSGlobals.MJS_FILE_EXTENSION)
+				&& !fileName.endsWith("." + N4JSGlobals.JSX_FILE_EXTENSION)) {
```
[the remark this answers is not in the document]
Comment (mor-n4) [reply testhelpers/org.eclipse.n4js.cli.tests.helper/src/org/eclipse/n4js/cli/helper/N4jscTestLanguageClient.java]: Done!

File: version.json (hunk 1 above)
Comment (mor-n4) [reply version.json:4]: Yes, the changes in the n4js-libs can have quite a far reaching impact.

Comment (mmews-n4) [reply version.json:4]: That means, from now on master will release tagged versions only until we decide otherwise. ok.

File: plugins/org.eclipse.n4js.transpiler.es/src/org/eclipse/n4js/transpiler/es/transform/CommonJsImportsTransformation.xtend (hunk 2 above)
[the remark this answers is not in the document]
Comment (mmews-n4) [reply plugins/org.eclipse.n4js.transpiler.es/src/org/eclipse/n4js/transpiler/es/transform/CommonJsImportsTransformation.xtend:81]: I see. How about adding a comment to `transformImportDecl` that it does the transformation in place and - additionally, will return further changes, like you said?

[the remark this answers is not in the document]
Comment (mor-n4) [reply plugins/org.eclipse.n4js.transpiler.es/src/org/eclipse/n4js/transpiler/es/transform/CommonJsImportsTransformation.xtend:88]: Good idea! Will do...

**Review by mmews-n4 — APPROVED** [review]
